fennel/expr/test_expr.py

```diff
@@ -0,0 +1,316 @@
+import pytest
+import pandas as pd
+from datetime import datetime
+from typing import Dict
+from fennel.datasets import dataset
+
+from fennel.expr import F, when
+from fennel.expr.visitor import ExprPrinter
+from fennel.expr.serializer import ExprSerializer
+from google.protobuf.json_format import ParseDict  # type: ignore
+from fennel.gen.expr_pb2 import Expr
+from fennel.testing.test_utils import error_message
+from fennel.expr.utils import compute_expr, compile
+
+
+def test_basic_expr1():
+    expr = (F("num") + F("d")).isnull()
+    df = pd.DataFrame({"num": [1, 2, 3, 4], "d": [5, 6, 7, 8]})
+    serializer = ExprSerializer()
+    print(compile(expr, {"num": int, "d": int}))
+    assert compile(expr, {"num": int, "d": int}) == bool
+    ret = compute_expr(expr, df, {"num": int, "d": int})
+    assert ret.tolist() == [False, False, False, False]
+
+
+def test_basic_expr2():
+
+    expr = F("a") + F("b") + 3
+    printer = ExprPrinter()
+    expected = "((Field[`a`] + Field[`b`]) + 3)"
+    assert expected == printer.print(expr.root)
+    serializer = ExprSerializer()
+    proto_expr = serializer.serialize(expr.root)
+    d = {
+        "binary": {
+            "left": {
+                "binary": {
+                    "left": {"ref": {"name": "a"}},
+                    "right": {"ref": {"name": "b"}},
+                }
+            },
+            "right": {
+                "jsonLiteral": {"literal": "3", "dtype": {"intType": {}}}
+            },
+        }
+    }
+    expected_expr = ParseDict(d, Expr())
+    assert expected_expr == proto_expr, error_message(proto_expr, expected_expr)
+
+    df = pd.DataFrame({"a": [1, 2, 3, 4], "b": [5, 6, 7, 8]})
+
+    @dataset
+    class TestDataset:
+        a: int
+        b: int
+        t: datetime
+
+    print(TestDataset.schema())
+    ret = compute_expr(expr, df, {"a": int, "b": int})
+    assert ret.tolist() == [9, 11, 13, 15]
+    ret = compute_expr(expr, df, TestDataset.schema())
+    assert ret.tolist() == [9, 11, 13, 15]
+    assert compile(expr, {"a": int, "b": int}) == int
+
+
+def test_math_expr():
+    expr = (F("a").num.floor() + 3.2).num.ceil()
+    printer = ExprPrinter()
+    expected = "CEIL((FLOOR(Field[`a`]) + 3.2))"
```

Comment:
Nit - let's print F('a') as Ref('a') vs Field. That way, it's going to be consistent with Rust backend + avoids confusion with dataset fields (since the same syntax needs to be reused for features later).